### PR TITLE
Add AnalyticData for NewtonianEuler

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

Different turbulence simulations require analytic initial data that are not solutions to the Euler equations. We currently don't have the directories for `PointwiseFunctions/AnalyticData/NewtonianEuler` so this PR is meant to add that support.

- Add directories for analytic data for NewtonianEuler
- Add initial data for Kelvin-Helmholtz instability. (This choice was arbitrary but simple + I've been using it quite a bit these past 2-3 weeks. Feel free to suggest a better alternative if you'd like.)

The PR is ~900 lines long but most of the code is just usual testing for pointwise functions.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsdeppe** [comment]
@carmaza the build is failing for what appears to be a legitimate reason :)

**Comment by carmaza** [comment]
OK thanks. I have no idea why that error has never shown up before, nor why I can't reproduce it locally. I pushed an extra commit adding the missing forward declaration. Hopefully that will fix it... 

I also squashed the fixup for the documentation.

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+ * \f}
+ *
+ * where \f$A\f$ is the amplitude, \f$\sigma\f$ is a damping factor, and
+ * \f$y_\text{top} = y_\text{mid} + b/2\f$ and
+ * \f$y_\text{bot} = y_\text{mid} - b/2\f$ are the vertical coordinates
+ * of the top and bottom boundaries of the strip, respectively.
+ *
+ * \note This class can be used to initialize 2D and 3D data. In 3D, the strip
+ * is aligned to the \f$x-\f$axis, and the vertical direction is taken to be
+ * the \f$z-\f$axis.
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp:97]: This note confused me: in 3D the strip should now be parallel to a plane, right? I don't understand what it means for it to be aligned to _one_ axis.

I think I would find this easier to understand if phrased as "In 3D, the strip is parallel to the xy-plane, and ..."

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+  using source_term_type = Sources::NoSource;
+
+  /// The adiabatic index of the fluid.
+  struct AdiabaticIndex {
+    using type = double;
+    static constexpr OptionString help = {"The adiabatic index of the fluid."};
+  };
+
+  /// The vertical coordinate of the horizontal bimedian of the strip.
+  struct StripBimedian {
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp]: [opinion:] `StripBimedian` might be more mathematically precise, but I think using something like `StripCenter` or `StripHeight` or `StripPosition` would be easier to understand... both in the name of this struct and for the associated variable names and documentation.

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    static constexpr OptionString help = {
+        "The thickness of the horizontal strip."};
+  };
+
+  /// The mass density in the strip
+  struct StripDensity {
+    using type = double;
+    static type lower_bound() noexcept { return 0.0; }
+    static constexpr OptionString help = {
+        "The mass density below the interface"};
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp]: Here and below, the help strings talk about "below the interface" and "above the interface", but the documentation makes no mention of interface and/or what it means to be above/below it. I would say "in the strip" and "outside of the strip" instead... this would better match the rho_in, rho_out in the documentation.

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.cpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      strip_half_thickness_(0.5 * strip_thickness),
+      strip_density_(strip_density),
+      strip_velocity_(strip_velocity),
+      background_density_(background_density),
+      background_velocity_(background_velocity),
+      pressure_(pressure),
+      perturbation_amplitude_(perturbation_amplitude),
+      damping_factor_(damping_factor),
+      equation_of_state_(adiabatic_index) {
+  ASSERT(strip_density_ > 0. and background_density_ > 0.,
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.cpp]: Elsewhere it looks like you've preferred `0.0` style

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+ * \f}
+ *
+ * where \f$b > 0\f$ is the thickness of the strip, and \f$y = y_\text{mid}\f$
+ * is its horizontal bimedian. The initial pressure is set equal to a constant,
+ * and the system is evolved assuming an ideal fluid of known adiabatic index.
+ * Finally, in order to excite the instability, the vertical velocity is
+ * initialized to a small perturbation
+ *
+ * \f{align*}
+ * v_y(x, y) = A\sin(4\pi x)
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp:65]: It looks like the periodic nature of the perturbation would have implications for the "allowed" sizes of the domain. Can these restrictions be described in the documentation (in a new paragraph)?

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  };
+
+  /// The amplitude of the perturbation
+  struct PerturbAmplitude {
+    using type = double;
+    static constexpr OptionString help = {"The amplitude of the perturbation."};
+  };
+
+  /// The damping factor of the perturbation
+  struct DampingFactor {
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp]: I would suggest something more descriptive, like `PerturbWidth`. The current `DampingFactor` doesn't tell me what is being damped and in what way...

**Review by fmahebert — COMMENTED** [review]
Some comments on documentation/naming, but the code itself looks good to me :)

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+#include "Utilities/TaggedTuple.hpp"
+
+/// \cond
+namespace PUP {
+class er;  // IWYU pragma: keep
+}  // namespace PUP
+/// \endcond
+
+namespace NewtonianEuler {
+namespace AnalyticData {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp:27]: Could you also add a new header `AnalyticData.hpp` in this directory that documents this namespace so one can navigate through `EvolutionSystems`->`NewtonianEuler`->`AnalyticData`?

File: tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+# See LICENSE.txt for details.
+
+import numpy as np
+
+
+def mass_density(x, adiabatic_index, strip_bimedian, strip_thickness,
+                 strip_density, strip_velocity, background_density,
+                 background_velocity, pressure, perturbation_amplitude,
+                 damping_factor):
+    return strip_density \
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.py]: [optional] I personally would find this code easier to read if instead of a ternary you did:
```python
if ...:
  return blah
return bla
```
and for `specific_internal_energy` too

File: tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+      "  DampingFactor: 0.01");
+  CHECK(kh_inst_from_options == kh_inst);
+
+  KhInstabilityProxy<Dim> kh_inst_to_move(
+      adiabatic_index, strip_bimedian, strip_thickness, strip_density,
+      strip_velocity, background_density, background_velocity, pressure,
+      perturbation_amplitude, damping_factor);
+  test_move_semantics(std::move(kh_inst_to_move), kh_inst);  //  NOLINT
+
+  test_serialization(kh_inst);
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp]: can you run the post-serialized state through the pypp test too?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks good, few minor comments. I'm happy for you to squash immediately

**Comment by carmaza** [comment]
Thanks for the review @fmahebert and @nilsdeppe I adopted all the suggestions and pushed individual fixup commits for easy review--I'll squash them upon approval.

File: tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp [hunk 10]
```diff
@@ -93,6 +92,14 @@ void test_analytic_data(const DataType& used_for_size) noexcept {
   test_move_semantics(std::move(kh_inst_to_move), kh_inst);  //  NOLINT
 
   test_serialization(kh_inst);
+
+  // run post-serialized state through checks with random numbers
+  pypp::check_with_random_values<
+      1, typename KhInstabilityProxy<Dim>::template variables_tags<DataType>>(
+      &KhInstabilityProxy<Dim>::template primitive_variables<DataType>, kh_inst,
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp]: Sorry, I meant `serailize_and_deserialize(kh_inst)` to make sure the serialization works correctly.

**Review by nilsdeppe — COMMENTED** [review]
I'm happy with you squashing, including this one change, once @fmahebert gives the go-ahead

File: tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp (hunk 10 above)
Comment (carmaza) [reply tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp]: Done, and I'm running it through pypp as well.

Comment (nilsdeppe) [reply tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp]: Sorry, I still wasn't clear enough :) I meant:
```cpp
pypp::check_with_random_values<
      1, typename KhInstabilityProxy<Dim>::template variables_tags<DataType>>(
      &KhInstabilityProxy<Dim>::template primitive_variables<DataType>, serialize_and_deserialize(kh_inst),
      "KhInstability",
      {"mass_density", "velocity", "specific_internal_energy", "pressure"},
      {{{0.0, 1.0}}}, members, used_for_size);
```
so that the deserialized version is checked. I should've been clearer initially, my apologies!

Comment (carmaza) [reply tests/Unit/PointwiseFunctions/AnalyticData/NewtonianEuler/Test_KhInstability.cpp]: Youre good. I thought the function mutated the argument instead of returning a new one.

**Review by nilsdeppe — COMMENTED** [review]
Okay, LGTM! :+1:

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp [hunk 11]
```diff
@@ -58,19 +59,39 @@ namespace AnalyticData {
...
+ * periodically along the \f$x-\f$axis. Here \f$A\f$ is the amplitude,
+ * \f$\sigma\f$ is a characteristic length for the perturbation width,
+ * and \f$y_\text{top} = y_\text{mid} + b/2\f$ and
  * \f$y_\text{bot} = y_\text{mid} - b/2\f$ are the vertical coordinates
  * of the top and bottom boundaries of the strip, respectively.
  *
+ * \note The periodic form of the perturbation enforces the horizontal
+ * extent of the domain to be a multiple of 0.5. The domain chosen in
+ * \cite Schaal2015 is the square \f$[0, 1]^2\f$, so the perturbation covers
+ * two wavelengths. In addition, the authors ran their simulations using
```
Comment (fmahebert) [inline src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp]: Since the wavelength is fixed and the domain size is not, I would rephrase this as: "..., so the perturbation covers two wavelengths." -> "..., which covers two wavelengths of the perturbation."

**Review by fmahebert — COMMENTED** [review]
Fixups look good to me, especially the expanded documentation! I have an optional minor rephrasing suggestion on one of the new sentences. You can squash the fixups and this new change.

File: src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp (hunk 11 above)
Comment (carmaza) [reply src/PointwiseFunctions/AnalyticData/NewtonianEuler/KhInstability.hpp]: Much better, yes. Thanks!

**Comment by carmaza** [comment]
Squashed.

**Review by fmahebert — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
image
